prplmesh-be: Add prplmesh_getall.lua script

VladyslavTupikin opened the pull request:
Need to add generic getall script for getting
data from UBus via ubus.sock and put it to the
mmx entrypoint.

Closes #1
Jira [PPM-1069](https://jira.prplfoundation.org/browse/PPM-1069) 
Signed-off-by: Vladyslav Tupikin <v.tupikin@inango-systems.com>

Commit:
prplmesh-be: Add prplmesh_getall.lua script

Need to add generic getall script for getting
data from UBus via ubus.sock and put it to the
mmx entrypoint.

Signed-off-by: Vladyslav Tupikin <v.tupikin@inango-systems.com>

MuteSpirit commented on src/prplmesh_getall.lua:
```diff
@@ -68,4 +68,184 @@
+
+function get_indexes(path)
+
+    local _ubus_connection
+    local devices
+    local instances
+    local result = {}
+
+    _ubus_connection = ubus.connect()
+    if not _ubus_connection then
+        print("Ubus sock conn fail")
+        return false
+    end
+
+    -- Retrieve output from list method via path from UBus
+    instance = _ubus_connection:call(path, "list", { })
+    if not instance then
+        print(path .. " not found in UBus")
+        return false
+    end
```
That's piece of code looks like common function `call_ubus(object, method, data)` from `prplmesh_get.lua` script
Move that function in separate Lua module and use it from all `prplmesh_*.lua` scripts

VladyslavTupikin replied:
deprecated commentary

MuteSpirit commented on src/prplmesh_getall.lua:
```diff
@@ -68,4 +68,184 @@
+    end
+
+    -- Fill up path instance indexes in table.
+    for k, v in pairs(instance["instances"]) do
```
If "instances" key will be absent in `instance` then you receive exception like
```
stdin:1: bad argument #1 to 'pairs' (table expected, got nil)
stack traceback:
	[C]: in function 'pairs'
	stdin:1: in main chunk
	[C]: ?
```
Please check "instance" presence before

VladyslavTupikin replied:
Fixed

MuteSpirit commented on src/prplmesh_getall.lua:
```diff
@@ -68,4 +68,184 @@
+        --print(" value: " .. tostring(result[k]))
+    end
+
+    _ubus_connection:close()
```
You should close ubus connection just after receiving alone result which you need
Because you have a `return` statement above this line and may receive non closed connection

VladyslavTupikin replied:
fixed

MuteSpirit replied:
What about `return` on line 101

MuteSpirit commented on src/prplmesh_getall.lua:
```diff
@@ -68,4 +68,184 @@
+
+function get_device_mmx_out(device_num)
+--TODO fix logic
+    local mmx_string = "0;"
```
Use a constant from `ing.ResCode` instead of hard-coded return code

VladyslavTupikin replied:
accepted

VladyslavTupikin replied:
fixed

MuteSpirit commented on src/prplmesh_getall.lua:
```diff
@@ -68,4 +68,184 @@
+
+function get_dm_table (path)
+
+    local tmp = {}
+    local parsed_args = {}
```
Create variables **after** verification input parameters
Otherwise they will be created but not used

VladyslavTupikin replied:
accepted

MuteSpirit commented on src/prplmesh_getall.lua:
```diff
@@ -68,4 +68,184 @@
+    end
+
+    for idx,val in pairs(tmp) do
+        if(idx == 1) then
```
You may not use parentheses here

VladyslavTupikin replied:
It is not true, we can use parentheses. but they are useless, removed.

MuteSpirit commented on src/prplmesh_getall.lua:
```diff
@@ -68,4 +68,184 @@
+            table.insert(tmp, str)
+    end
+
+    for idx,val in pairs(tmp) do
```
Use one space after comma
```Lua
for idx, val in pairs(tmp) do
```

VladyslavTupikin replied:
accepted

MuteSpirit commented on src/prplmesh_getall.lua:
```diff
@@ -68,4 +68,184 @@
+        return false
+    end
+
+    for str in string.gmatch(path, "(.-){i}") do
+            table.insert(tmp, str)
+    end
```
What do you think about idea to match both indexes and sub-paths of objects by single loop:
```
> s = "Controller.Network.Device.{i}.Radio.{i}.BSS.{i}.STA.{i}."
> r={}
> for m in string.gmatch(s, "(.-)[.]{i}[.]?") do print(m); table.insert(r, m); end
Controller.Network.Device
Radio
BSS
STA
> for k,v in ipairs(r) do print("r["..tostring(k).."] = "..tostring(v)); end
r[1] = Controller.Network.Device
r[2] = Radio
r[3] = BSS
r[4] = STA
```

VladyslavTupikin replied:
Need to test. If it is nitpick i prefer to skip it.

VladyslavTupikin replied:
Accepted!

MuteSpirit commented on src/prplmesh_getall.lua:
```diff
@@ -68,4 +68,315 @@
+
+    for idx,val in pairs(tmp) do
+        if(idx == 1) then
+            parsed_args[idx] = tostring(val):sub(1, #tostring(val) - 1)
```
You may use `sub(1, -2)` instead of `#tostring(val) - 1`

VladyslavTupikin replied:
Accepted

MuteSpirit commented on src/prplmesh_getall.lua:
```diff
@@ -68,4 +68,414 @@
+    -- Retrieve output from list method via path from UBus
+    instance = _ubus_connection:call(path, "list", { })
+    if not instance then
+        print(path .. " not found in UBus")
```
If you don't want to break MMX EP parsing of script results then print your error/debug messages into `stderr`
See 
    https://github.com/InangoSystems/mmx-ep/blob/033b05fef89c1ea63ba3f87efcf90b601ee74560/src/ep_worker.c#L974
    https://github.com/InangoSystems/mmx-ep/blob/033b05fef89c1ea63ba3f87efcf90b601ee74560/src/ep_worker.c#L1033
    for details about MMX EP behavior

VladyslavTupikin replied:
added new function `error()` and replaced all prints by this function.

MuteSpirit commented on src/prplmesh_getall.lua:
```diff
@@ -68,4 +68,414 @@
+
+
+--[[
+    @brief  Function for convrerting path to table.
```
Incorrect description
Function splits object path string into list of root and sub-objects

VladyslavTupikin replied:
Fixed

MuteSpirit commented on src/prplmesh_getall.lua:
```diff
@@ -68,4 +68,414 @@
+        return false
+    end
+
+    local tmp = {}
```
unused. delete

VladyslavTupikin replied:
removed

MuteSpirit commented on src/prplmesh_getall.lua:
```diff
@@ -68,4 +68,414 @@
+--]]
+local function write_mmx_out(path, data)
+
+    local file = io.open(path, "a+")
```
Why is it needed to format script output string in separate file instead of script variable

VladyslavTupikin replied:
solved

MuteSpirit commented on src/prplmesh_getall.lua:
```diff
@@ -68,4 +68,414 @@
+        return false
+    end
+
+    root = {
```
Is really `root` object is different in comparison to all other nodes and require separate function for creation?

MuteSpirit commented on src/prplmesh_getall.lua:
```diff
@@ -68,4 +68,414 @@
+local function get_mmx_out(path)
+
+
+    local mmx_out_file = "/tmp/getall"
```
MMX EP has more then 1 working thread and able to run more then 1 getall script simulaneously
That's one more reason not use file to store the output result string

VladyslavTupikin replied:
solved

MuteSpirit commented on src/prplmesh_getall.lua:
```diff
@@ -68,4 +68,414 @@
+    @param  path Path to file.
+--]]
+local function dfs(root, path)
+    if root then
```
The next error handling style make function shorter:
```
if not root then
    return
end

another code
```

MuteSpirit commented on src/prplmesh_getall.lua:
```diff
@@ -68,4 +68,414 @@
+--]]
+local function dfs(root, path)
+    if root then
+        for count, idx in pairs(root.idx_tbl) do
```
Why not iterate list of child nodes (`root.k`) instead of mark then `visited` or not?
Do you even check value of `visited` field?

VladyslavTupikin replied:
I am checking this value for last object for avoiding result duplication.

MuteSpirit reviewed (changes requested):
By my opinion this script may be written more optimal
For example, if make tree in view of
```
treeRoot =
{
    path = nil,
    nodes = {
        [1] = {
            path="A.1",
            nodes=
            {
                [1] = { 
                    path="A.1.B.C.1",
                    nodes=nil,
                },
                [2] = { 
                    path="A.1.B.C.2",
                    nodes=nil,
                },
            },
        },
        [2] = {
            path="A.2",
            nodes={
                [1] = { 
                    path="A.2.B.C.1",
                    nodes=nil,
                },
            },
        }
    }
}
```

Then next function will be able to collect all needed indexes:
```
function collect_list_indexes(nodeIndexes, node, resIndexes)
    if not node["nodes"] then
        table.insert(resIndexes, nodeIndexes)
        return
    end

    for idx, child in ipairs(node["nodes"]) do
        local childIndexes = copyTable(nodeIndexes)
        table.insert(childIndexes, idx)

        collect_list_indexes(childIndexes, child, resIndexes)
    done
end


res = {}
treeRoot

collect_list_indexes({}, treeRoot, res)

res should be {{1, 1},{1, 2},{2, 1}}
```

Commit:
prplmesh-be: Print error messages into stdetr

Added function for printing error messages into stderr.

Fixed description for get_dm_table().

Also added few changes in conditions because sometimes
null comes there instead of proper table.

Signed-off-by: Vladyslav Tupikin <v.tupikin@inango-systems.com>

Commit:
prplmesh-be: Replace file usage to global variable

In prplmesh_getall.lua replace file usage to
global variable for saving mmx result string,
remove functions for working with files.

Signed-off-by: Vladyslav Tupikin <v.tupikin@inango-systems.com>
